ribbon-eureka/src/main/java/com/netflix/loadbalancer/EurekaDynamicNotificationLoadBalancer.java

```diff
@@ -0,0 +1,159 @@
+                    .setNameFormat("EurekaDynamicNotificationLoadBalancer-ServerListUpdate-%d")
+                    .setDaemon(true)
+                    .build()
+    );
+
+    private final EurekaEventListener updateListener = new EurekaEventListener() {
+        @Override
+        public void onEvent(EurekaEvent event) {
+            if (event instanceof CacheRefreshedEvent) {
+                serverListUpdateService.submit(new Runnable() {  // fire and forget
+                    @Override
+                    public void run() {
+                        EurekaDynamicNotificationLoadBalancer.super.updateListOfServers();
+                    }
+                });
+            }
+        }
+    };
+
+    private final Provider<EurekaClient> eurekaClientProvider;
+    private final ExecutorService serverListUpdateService;
+
+    // necessary for legacy use cases :(
+    @Deprecated
+    public EurekaDynamicNotificationLoadBalancer() {
+        this(
+                null,
+                new LegacyEurekaClientProvider(),
+                DEFAULT_SERVER_LIST_UPDATE_EXECUTOR
+        );
+    }
+
+    public EurekaDynamicNotificationLoadBalancer(IClientConfig clientConfig,
+                                                 Provider<EurekaClient> eurekaClientProvider,
+                                                 ExecutorService serverListUpdateService) {
+        super();
+
+        this.eurekaClientProvider = eurekaClientProvider;
+        this.serverListUpdateService = serverListUpdateService;
+        if (clientConfig != null) {
+            initWithNiwsConfig(clientConfig);
+        }
+    }
+
+    @Override
+    public void initWithNiwsConfig(IClientConfig clientConfig) {
+        try {
+            super.initWithNiwsConfig(clientConfig);
+
+            DiscoveryEnabledNIWSServerList myServerList = new DiscoveryEnabledNIWSServerList(
+                    clientConfig,
+                    eurekaClientProvider
+            );
+            setServerListImpl(myServerList);
+
+            AbstractServerListFilter<DiscoveryEnabledServer> myFilter = myServerList.getFilterImpl(clientConfig);
+            myFilter.setLoadBalancerStats(getLoadBalancerStats());
+            setFilter(myFilter);
+
+            connectToEureka();
+
+            LOGGER.info("EurekaDynamicNotificationLoadBalancer for client {} initialized: {}",
+                    clientConfig.getClientName(), this.toString());
+        } catch (Exception e) {
+            throw new RuntimeException("Exception while initializing EurekaDynamicNotificationLoadBalancer," +
+                    " niwsClientConfig: " + clientConfig, e);
+        }
+    }
+
+    /**
+     * Makes no sense to ping an inmemory disc client
+     */
+    @Override
+    public void forceQuickPing() {
+        // no-op
+    }
+
+    @Override
+    public void stopServerListRefreshing() {
+        if (eurekaClientProvider != null) {
+            EurekaClient eurekaClient = eurekaClientProvider.get();
```

Comment:
If there is an error starting eureka client in the constructor this will end up duplicating the operation.  You may want instead to store the eurekaClient and provider in separate variables